### PR TITLE
feature/register

### PR DESCRIPTION
**Pull request by msadovyi**

### CODE REVIEW

File: packages/backend/src/modules/user/user.controller.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    private readonly logger: LoggerService,
+  ) { }
+
+  @Get('/confirm-email')
+  @Redirect('', 302)
+  public async confirmEmail(@Query('email') email: string, @Query('code') code: string): Promise<{ url: string }> {
+    try {
+      const result = await this.service.acceptConfirmationCode(email, ConfirmationCodeTypes.EMAIL_CONFIRMATION, code);
+      if (!result) {
+        throw new Error('Can not confirm email address');
```
Comment (valcker) [inline packages/backend/src/modules/user/user.controller.ts:26]: Cannot

File: packages/backend/src/modules/user/user.service.ts [hunk 2]
```diff
@@ -27,22 +34,28 @@ export default class UserService {
   public async login(email: string, password: string, device: string): Promise<ITokens> {
     const user = await this.userModel.findOne({ email });
 
+    if ((await this.isEmailNotConfirmed(email))) {
+      await this.sendEmailConfirmationCode(email);
+      throw new AuthenticationError('We have sent a confirmation to you. Confirm your email address, please.');
```
Comment (valcker) [inline packages/backend/src/modules/user/user.service.ts:39]: Please verify your email by using a confirmation code we have sent to your email.

File: packages/backend/src/modules/user/user.service.ts [hunk 3]
```diff
@@ -70,4 +83,107 @@ export default class UserService {
       _id: { $ne: exceptUser },
     });
   }
+
+  public async resetPasswordRequest(email: string): Promise<boolean> {
+    const user = await this.findByEmail(email);
+    if (!user) {
+      throw new UserInputError(`User with this email does not exist`);
```
Comment (valcker) [inline packages/backend/src/modules/user/user.service.ts:90]: This is information disclosure which a potential attacker can use to narrow down the brute-force attack. Replace with "User with such email/password does not exist".

File: packages/backend/src/modules/user/user.service.ts [hunk 4]
```diff
@@ -70,4 +83,107 @@ export default class UserService {
...
+        }
+      }
+    );
+
+    return true;
+  }
+
+  public async resetPassword(email: string, code: string, password: string): Promise<boolean> {
+    const isConfirmed = await this.acceptConfirmationCode(email, CodeTypes.RESET_PASSWORD, code);
+    if (!isConfirmed) throw new UserInputError('Confirmation code was not found');
```
Comment (valcker) [inline packages/backend/src/modules/user/user.service.ts:112]: Confirmation code has expired or is invalid

File: packages/backend/src/modules/user/user.service.ts [hunk 5]
```diff
@@ -70,4 +83,107 @@ export default class UserService {
...
+
+    const user = await this.findByEmail(email);
+
+    if (!user) {
+      throw new UserInputError(`User with email: ${email} was found`);
```
Comment (valcker) [inline packages/backend/src/modules/user/user.service.ts:117]: Same as above, replace the message with "User with such email/password does not exist".

File: packages/backend/src/modules/user/user.service.ts [hunk 6]
```diff
@@ -70,4 +83,107 @@ export default class UserService {
...
+
+    await user.save();
+
+    return true;
+  }
+
+  public async sendEmailConfirmationCode(email: string): Promise<void> {
+    const user = await this.findByEmail(email);
+    if (!user) {
+      throw new UserInputError(`User with this email does not exist`);
```
Comment (valcker) [inline packages/backend/src/modules/user/user.service.ts:129]: Replace the message with "User with such email/password does not exist".

File: packages/backend/src/modules/user/user.controller.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      }
+      return {
+        url: `${config.client.host}/confirm-email?result=success`
+      };
+    } catch (error) {
```
Comment (valcker) [inline packages/backend/src/modules/user/user.controller.ts:31]: Do we really need try/catch here? Why not just use if/else -> return?

File: packages/backend/src/modules/user/user.service.ts [hunk 8]
```diff
@@ -70,4 +83,107 @@ export default class UserService {
...
+    const context = `${this.constructor.name}:${this.acceptConfirmationCode.name}`;
+    this.logger.debug('Looking for confirmation code', context, { email, code, type });
+    const confirmationCode = await this.codeModel.findOne({
+      code,
+      email,
+      type,
+      expiredAt: {
+        $gte: new Date(),
+      }
+    });
```
Comment (valcker) [inline packages/backend/src/modules/user/user.service.ts:178]: ESLint warning, semicolon is not needed.

File: packages/frontend/src/presenter/ResetPassword/useResetPassword.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  password: string;
+  confirmPassword: string;
+};
+
+const validationSuite = vest.create('reset_password_request_form', ({ password, confirmPassword }: ResetPasswordFormInput) => {
+  test('password', 'Password is required', () => {
+    enforce(password).isNotEmpty();
+  });
+  test('password', 'Password must be at least 5 chars', () => {
+    enforce(password).longerThanOrEquals(5);
```
Comment (valcker) [inline packages/frontend/src/presenter/ResetPassword/useResetPassword.ts:20]: Let's raise to 8 chars, please.

File: packages/frontend/src/view/ConfirmEmail/ConfirmEmail.tsx [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+export const ConfirmEmail: React.FC = () => {
+  const { content, logo } = useStyles();
+  const { location } = useHistory();
+  const { result, error } = getQueryData(location.search);
+  const { accessToken } = useAuth();
+
+  if (accessToken) return <Redirect to={ROUTES.MAIN} />
+
+  const message = result === 'success' ? <H3> You email was successfully confirmed! 🎉 </H3> : <H3>  😢 {error || 'Oops... Something went wrong. Try again later, please'} </H3>
```
Comment (valcker) [inline packages/frontend/src/view/ConfirmEmail/ConfirmEmail.tsx:19]: _Your_ email...

File: packages/backend/src/modules/user/user.service.ts (hunk 8 above)
Comment (msadovyi) [reply packages/backend/src/modules/user/user.service.ts:178]: I don't see any warnings, prettier puts it here as it the end of statement.

File: packages/frontend/src/view/ResetPassword/useStyles.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    width: "100%",
+    flexDirection: 'column',
+  },
+  links: {
+    marginTop: spacing(2),
+    alignSelf: 'center',
+    '& a': {
+      fontWeight: 'bold',
+      textDecoration: 'none',
+      color: palette.common.white,
```
Comment (Feshchenko) [inline packages/frontend/src/view/ResetPassword/useStyles.ts:30]: `color: palette.text.primary`

File: packages/frontend/src/view/SignUp/useStyles.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    width: "100%",
+    flexDirection: 'column',
+  },
+  links: {
+    marginTop: spacing(2),
+    alignSelf: 'center',
+    '& a': {
+      fontWeight: 'bold',
+      textDecoration: 'none',
+      color: palette.common.white,
```
Comment (Feshchenko) [inline packages/frontend/src/view/SignUp/useStyles.ts:30]: `color: palette.text.primary`

File: packages/backend/src/modules/user/user.service.ts (hunk 8 above)
Comment (valcker) [reply packages/backend/src/modules/user/user.service.ts:178]: Bizarre, the comment is on the wrong line for some reason. I'll close this one and will add another one.

File: packages/backend/src/modules/user/user.service.ts [hunk 13]
```diff
@@ -70,4 +83,107 @@ export default class UserService {
...
+    if (!confirmationCode) {
+      this.logger.debug('Confirmation code was not found', context, { email, code });
+      return false;
+    };
```
Comment (valcker) [inline packages/backend/src/modules/user/user.service.ts:182]: here, the semicolon is not needed here.

File: packages/backend/src/modules/user/user.service.ts [hunk 14]
```diff
@@ -27,22 +34,28 @@ export default class UserService {
   public async login(email: string, password: string, device: string): Promise<ITokens> {
     const user = await this.userModel.findOne({ email });
 
+    if ((await this.isEmailNotConfirmed(email))) {
+      await this.sendEmailConfirmationCode(email);
+      throw new AuthenticationError('We have sent a confirmation to you. Confirm your email address, please.');
+    };
```
Comment (valcker) [inline packages/backend/src/modules/user/user.service.ts:40]: same here, no ";" after the if.

**Review by valcker — APPROVED** [review]

**Pull request merged**